
This fixes DRAM size detection on H313 DRAM controller.

All kudos to junari who identifes problem and proposed fix


diff --speed-large-files --no-dereference --minimal -Naur u-boot-sunxi-647b392bf20614006917d5fcd3390347a668eee5/arch/arm/mach-sunxi/dram_sun50i_h616.c u-boot-sunxi-647b392bf20614006917d5fcd3390347a668eee5/arch/arm/mach-sunxi/dram_sun50i_h616.c
--- u-boot-sunxi-647b392bf20614006917d5fcd3390347a668eee5/arch/arm/mach-sunxi/dram_sun50i_h616.c	2023-04-07 12:26:13.844887697 +0200
+++ u-boot-sunxi-647b392bf20614006917d5fcd3390347a668eee5/arch/arm/mach-sunxi/dram_sun50i_h616.c	2023-04-07 12:25:23.011556227 +0200
@@ -833,20 +833,20 @@
 	clrsetbits_le32(SUNXI_DRAM_PHY0_BASE + 0x3c, 0xf, val);
 
 	if (para->type == SUNXI_DRAM_TYPE_DDR3) {
-		writel(0xd, SUNXI_DRAM_PHY0_BASE + 0x14);
-		writel(0xd, SUNXI_DRAM_PHY0_BASE + 0x35c);
-		writel(0xd, SUNXI_DRAM_PHY0_BASE + 0x368);
-		writel(0xd, SUNXI_DRAM_PHY0_BASE + 0x374);
+		writel(0x9, SUNXI_DRAM_PHY0_BASE + 0x14);
+		writel(0x9, SUNXI_DRAM_PHY0_BASE + 0x35c);
+		writel(0x9, SUNXI_DRAM_PHY0_BASE + 0x368);
+		writel(0x9, SUNXI_DRAM_PHY0_BASE + 0x374);
 
 		writel(0, SUNXI_DRAM_PHY0_BASE + 0x18);
 		writel(0, SUNXI_DRAM_PHY0_BASE + 0x360);
 		writel(0, SUNXI_DRAM_PHY0_BASE + 0x36c);
 		writel(0, SUNXI_DRAM_PHY0_BASE + 0x378);
 
-		writel(9, SUNXI_DRAM_PHY0_BASE + 0x1c);
-		writel(9, SUNXI_DRAM_PHY0_BASE + 0x364);
-		writel(9, SUNXI_DRAM_PHY0_BASE + 0x370);
-		writel(9, SUNXI_DRAM_PHY0_BASE + 0x37c);
+		writel(7, SUNXI_DRAM_PHY0_BASE + 0x1c);
+		writel(7, SUNXI_DRAM_PHY0_BASE + 0x364);
+		writel(7, SUNXI_DRAM_PHY0_BASE + 0x370);
+		writel(7, SUNXI_DRAM_PHY0_BASE + 0x37c);
 	}
 	if (para->type == SUNXI_DRAM_TYPE_LPDDR3) {
 		writel(0xe, SUNXI_DRAM_PHY0_BASE + 0x14);
@@ -1005,7 +1005,7 @@
 	 * specified below.
 	 */
 	if (para->type == SUNXI_DRAM_TYPE_DDR3) {
-		writel(0x1f14, &mctl_ctl->mrctrl1);
+		writel(0x1b50, &mctl_ctl->mrctrl1);
 		writel(0x80000030, &mctl_ctl->mrctrl0);
 		mctl_await_completion(&mctl_ctl->mrctrl0, BIT(31), 0);
 
@@ -1013,7 +1013,7 @@
 		writel(0x80001030, &mctl_ctl->mrctrl0);
 		mctl_await_completion(&mctl_ctl->mrctrl0, BIT(31), 0);
 
-		writel(0x20, &mctl_ctl->mrctrl1);
+		writel(0x10, &mctl_ctl->mrctrl1);
 		writel(0x80002030, &mctl_ctl->mrctrl0);
 		mctl_await_completion(&mctl_ctl->mrctrl0, BIT(31), 0);
 
diff --speed-large-files --no-dereference --minimal -Naur u-boot-sunxi-647b392bf20614006917d5fcd3390347a668eee5/arch/arm/mach-sunxi/dram_timings/h616_ddr3_1333.c u-boot-sunxi-647b392bf20614006917d5fcd3390347a668eee5/arch/arm/mach-sunxi/dram_timings/h616_ddr3_1333.c
--- u-boot-sunxi-647b392bf20614006917d5fcd3390347a668eee5/arch/arm/mach-sunxi/dram_timings/h616_ddr3_1333.c	2021-04-26 11:12:35.000000000 +0200
+++ u-boot-sunxi-647b392bf20614006917d5fcd3390347a668eee5/arch/arm/mach-sunxi/dram_timings/h616_ddr3_1333.c	2023-04-07 12:26:00.334888195 +0200
@@ -85,8 +85,9 @@
 	clrsetbits_le32(&mctl_ctl->rankctl, 0xff0, 0x660);
 
 	/* Configure DFI timing */
-	writel((tcl - 2) | 0x2000000 | (t_rdata_en << 16) | 0x808000,
-	       &mctl_ctl->dfitmg0);
+	//writel((tcl - 2) | 0x2000000 | (t_rdata_en << 16) | 0x808000,
+	//       &mctl_ctl->dfitmg0);
+	writel(0x2858003, &mctl_ctl->dfitmg0);
 	writel(0x100202, &mctl_ctl->dfitmg1);
 
 	/* set refresh timing */
